File: payout_bot.py
import os
import time
import re
import base64
import logging
from decimal import Decimal
from dotenv import load_dotenv

# Cài đặt Logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s',
    handlers=[
        logging.FileHandler("bot.log"),
        logging.StreamHandler()
    ]
)

try:
    from pysui import SuiConfig, SyncClient
    from pysui.sui.sui_builders.get_builders import QueryTransactionBlocks
    from pysui.sui.sui_builders.exec_builders import PaySui
    from pysui.sui.sui_types.scalars import SuiAddress
    from pysui.sui.sui_crypto import Ed25519KeyPair
    from pysui.sui.sui_txresults.complex_txresults import TransactionBlock
except ImportError:
    logging.error("Missing libraries. Please run: pip install pysui python-dotenv")
    exit(1)

# --- CONFIGURATION ---
load_dotenv()
SECRET_KEY_B64 = os.getenv('SUI_HOUSE_SECRET_KEY_B64')
HOUSE_ADDRESS = os.getenv('SUI_HOUSE_ADDRESS')
NETWORK = os.getenv('SUI_NETWORK', 'testnet')
CHECK_INTERVAL = 10  # Seconds
DB_FILE = "processed_digests.txt"

if not SECRET_KEY_B64 or not HOUSE_ADDRESS:
    logging.error("❌ SUI_HOUSE_SECRET_KEY_B64 or SUI_HOUSE_ADDRESS is missing in .env")
    exit(1)

# Khởi tạo Keypair từ Base64
def get_keypair(b64_str):
    data = base64.b64decode(b64_str)
    # Nếu là 33 bytes (có flag Ed25519 ở đầu), bỏ byte đầu tiên
    if len(data) == 33 and data[0] == 0:
        data = data[1:]
    return Ed25519KeyPair.from_bytes(data)

keypair = get_keypair(SECRET_KEY_B64)
# pysui config tự động dựa trên network
# Đối với bot độc lập, ta có thể dùng config mặc định hoặc tạo mới
try:
    config = SuiConfig.default_config()
except Exception:
    # Nếu chưa có config (chưa chạy sui client), ta tạo config tối thiểu
    logging.warning("No default Sui config found. Using manual config.")
    # Lưu ý: Pysui thường yêu cầu config file. Trên VPS bạn nên chạy `sui client` trước 1 lần.
    # Hoặc ta có thể dùng SyncClient với URL trực tiếp nếu pysui hỗ trợ.
    config = SuiConfig.user_config(rpc_url="https://fullnode.testnet.sui.io:443")

client = SyncClient(config)

# Lưu trữ các giao dịch đã xử lý vào file
def load_processed():
    if not os.path.exists(DB_FILE):
        return set()
    with open(DB_FILE, "r") as f:
        return set(line.strip() for line in f if line.strip())

def save_processed(digest):
    with open(DB_FILE, "a") as f:
        f.write(f"{digest}\n")

processed_digests = load_processed()

def get_suffix(amount_mist):
    """Lấy số cuối cùng của phần thập phân sau khi chia 10^9 (giống JS)"""
    sui_amount = Decimal(amount_mist) / Decimal(10**9)
    sui_str = format(sui_amount.normalize(), 'f')
    if "." not in sui_str:
        return None
    decimal_part = sui_str.split(".")[1]
    if not decimal_part:
        return None
    return decimal_part[-1]

def check_and_payout():
    global processed_digests
    logging.info(f"🔍 Scanning transactions for {HOUSE_ADDRESS}...")
    
    try:
        # Query các giao dịch đến địa chỉ nhà cái
        builder = QueryTransactionBlocks(
            query={"ToAddress": HOUSE_ADDRESS},
            descending_order=True,
            limit=20
        )
        
        # Thêm options để xem balanceChanges và input
        # Lưu ý: Pysui QueryTransactionBlocks có thể cần tham số khác tùy phiên bản
        # Ở đây ta giả định cấu trúc chuẩn của Sui RPC
        result = client.execute(builder)
        
        if not result.is_ok():
            logging.error(f"Error querying transactions: {result.result_string}")
            return

        tx_blocks = result.result_data.data
        
        for tx in tx_blocks:
            digest = tx.digest
            if digest in processed_digests:
                continue
            
            sender = tx.transaction.data.sender
            
            # Kiểm tra số tiền nhận được (balanceChanges)
            # Lưu ý: Cấu trúc dữ liệu của pysui có thể bọc trong các Object
            house_change = None
            if hasattr(tx, 'balance_changes'):
                for change in tx.balance_changes:
                    if change.owner.get('AddressOwner') == HOUSE_ADDRESS:
                        amount = int(change.amount)
                        if amount > 0:
                            house_change = amount
                            break
            
            if not house_change:
                processed_digests.add(digest)
                continue

            amount_mist = house_change
            suffix = get_suffix(amount_mist)
            
            if not suffix:
                processed_digests.add(digest)
                continue
                
            # Lấy số cuối của Digest
            digits = re.findall(r'\d', digest)
            if not digits:
                processed_digests.add(digest)
                continue
            last_digit = int(digits[-1])
            
            is_win = False
            ratio = 0
            
            # Logic game
            if suffix == '1' and last_digit in [1, 3, 5, 7]:
                is_win = True; ratio = 2.4
            elif suffix == '2' and last_digit in [2, 4, 6, 8]:
                is_win = True; ratio = 2.4
            elif suffix == '4' and last_digit in [5, 6, 7, 8]:
                is_win = True; ratio = 2.2
            elif suffix == '3' and last_digit in [1, 2, 3, 4]:
                is_win = True; ratio = 2.2
                
            if is_win:
                reward_amount = int(amount_mist * ratio)
                logging.info(f"✅ WINNER: {sender} won {reward_amount/1e9} SUI (Bet: {amount_mist/1e9}, Ratio: {ratio})")
                
                # Gửi Payout
                payout_tx = PaySui(
                    recipients=[SuiAddress(sender)],
                    amounts=[reward_amount]
                )
                
                # Thực hiện ký và gửi
                payout_result = client.execute(payout_tx, keypair)
                
                if payout_result.is_ok():
                    logging.info(f"💸 Payout sent! Digest: {payout_result.result_data.digest}")
                else:
                    logging.error(f"❌ Payout failed for {digest}: {payout_result.result_string}")
                    # Nếu lỗi do gas hoặc network, có thể thử lại sau (không add vào processed)
                    continue

            # Đánh dấu đã xử lý
            processed_digests.add(digest)
            save_processed(digest)
            
    except Exception as e:
        logging.error(f"Error in main loop: {str(e)}")

if __name__ == "__main__":
    logging.info("🤖 Sui Payout Bot (Python) started...")
    logging.info(f"🌍 Network: {NETWORK}")
    
    while True:
        check_and_payout()
        time.sleep(CHECK_INTERVAL)

File: payout-bot.js
import 'dotenv/config';
import { SuiJsonRpcClient as SuiClient, getJsonRpcFullnodeUrl as getFullnodeUrl } from '@mysten/sui/jsonRpc';
import { Ed25519Keypair } from '@mysten/sui/keypairs/ed25519';
import { Transaction } from '@mysten/sui/transactions';
import { fromBase64 } from '@mysten/sui/utils';

/**
 * 💡 LƯU Ý QUAN TRỌNG VỀ BẢO MẬT:
 * 1. Bot này dùng biến môi trường trong file .env để bảo mật.
 * 2. Đảm bảo file .env của bạn đã có SUI_HOUSE_SECRET_KEY_B64 và SUI_HOUSE_ADDRESS.
 */

// --- CONFIGURATION ---
const SECRET_KEY_B64 = process.env.SUI_HOUSE_SECRET_KEY_B64; 
const HOUSE_ADDRESS = process.env.SUI_HOUSE_ADDRESS;
const NETWORK = process.env.SUI_NETWORK || 'testnet';
// ---------------------

if (!SECRET_KEY_B64 || !HOUSE_ADDRESS) {
    console.error('❌ Error: SUI_HOUSE_SECRET_KEY_B64 or SUI_HOUSE_ADDRESS is missing in .env');
    process.exit(1);
}

const client = new SuiClient({ url: getFullnodeUrl(NETWORK) });

// Xử lý Secret Key nếu có flag (33 bytes)
let secretKey = fromBase64(SECRET_KEY_B64);
if (secretKey.length === 33 && secretKey[0] === 0) {
    secretKey = secretKey.slice(1);
}
const keypair = Ed25519Keypair.fromSecretKey(secretKey);

console.log('🤖 Payout Bot is starting...');
console.log('🏠 Monitoring House:', HOUSE_ADDRESS);

// Tập hợp các digest đã xử lý để tránh trả thưởng trùng
const processedDigests = new Set();

async function checkAndPayout() {
    try {
        const txs = await client.queryTransactionBlocks({
            filter: { ToAddress: HOUSE_ADDRESS },
            options: { showBalanceChanges: true, showInput: true, showEffects: true },
            limit: 20,
            descendingOrder: true
        });

        for (const tx of txs.data) {
            const digest = tx.digest;
            
            // Nếu đã xử lý rồi thì bỏ qua
            if (processedDigests.has(digest)) continue;

            const sender = tx.transaction.data.sender;
            
            // Tìm sự thay đổi số dư của nhà cái (số dương là tiền nhận được)
            const houseChange = tx.balanceChanges.find(bc => bc.owner.AddressOwner === HOUSE_ADDRESS);
            if (!houseChange || parseInt(houseChange.amount) <= 0) continue;

            const amountMist = parseInt(houseChange.amount);
            const rawAmountStr = (amountMist / 1000000000).toString();
            
            // Logic cược:
            const decimalPartMatch = rawAmountStr.split('.')[1];
            if (!decimalPartMatch) continue;
            
            const decimalPart = decimalPartMatch.replace(/0+$/, "");
            if (decimalPart.length === 0) continue;
            
            const suffix = decimalPart.charAt(decimalPart.length - 1);
            const digits = digest.match(/\d/g);
            if (!digits) continue;
            const lastDigit = parseInt(digits[digits.length - 1]);

            let isWin = false;
            let ratio = 0;

            if (suffix === '1' && [1, 3, 5, 7].includes(lastDigit)) { isWin = true; ratio = 2.4; }
            else if (suffix === '2' && [2, 4, 6, 8].includes(lastDigit)) { isWin = true; ratio = 2.4; }
            else if (suffix === '4' && [5, 6, 7, 8].includes(lastDigit)) { isWin = true; ratio = 2.2; }
            else if (suffix === '3' && [1, 2, 3, 4].includes(lastDigit)) { isWin = true; ratio = 2.2; }

            if (isWin) {
                const rewardAmount = Math.floor(amountMist * ratio);
                console.log(`✅ WINNER detected! Sending ${rewardAmount / 1e9} SUI to ${sender}`);
                await sendPayout(sender, rewardAmount);
            }
            
            // Đánh dấu là đã xử lý
            processedDigests.has(digest) || processedDigests.add(digest);
        }
    } catch (e) {
        console.error('Error in bot loop:', e.message);
    }
}

async function sendPayout(recipient, amount) {
    const tx = new Transaction();
    const [coin] = tx.splitCoins(tx.gas, [tx.pure.u64(amount)]);
    tx.transferObjects([coin], tx.pure.address(recipient));

    try {
        const result = await client.signAndExecuteTransaction({
            signer: keypair,
            transaction: tx,
        });
        console.log('💸 Payout sent! Digest:', result.digest);
    } catch (e) {
        console.error('Failed to send payout:', e.message);
    }
}

setInterval(checkAndPayout, 10000);
checkAndPayout();
